src/librustc_trans/trans/_match.rs

```diff
@@ -630,8 +630,7 @@ fn bind_subslice_pat(bcx: Block,
     let vec_datum = match_datum(val, vec_ty);
     let (base, len) = vec_datum.get_vec_base_and_len(bcx);
 
-    let slice_byte_offset = Mul(bcx, vt.llunit_size, C_uint(bcx.ccx(), offset_left));
-    let slice_begin = tvec::pointer_add_byte(bcx, base, slice_byte_offset);
+    let slice_begin = InBoundsGEP(bcx, base, &[C_uint(bcx.ccx(), offset_left)]);
```

Comment:
@tbu- it's likely that we can regard any pointer-to-zero-sized-type as valid, but that doesn't mean that it is valid/sensible for this specific pointer to move outside the object from which it supposedly came; for one, how we did it was UB, as @dotdash said, and, Rust code can care about the actual real addresses of pointers so this is incorrect in that respect.